strawberryfields/backends/states.py

```diff
@@ -814,6 +815,43 @@ def expand_dims(op, n, modes):
 
         return mean, var
 
+    def number_expectation(self, modes):
+        """
+        Calculates the expectation value of the product of the number operators of the modes.
+        Args:
+            modes (list): list of modes for which one wants the expectation of the product of their number operator.
+        Return:
+            (float): the expectation value.
+        """
+        if len(modes) != len(set(modes)):
+            raise ValueError("there can be no duplicates in modes")
+
+
+        state = self._data #representation of the quantum state either as tensor of probability amplitudes
+        # or as
+        pure = self._pure #if pure or as a tensor of density matrix elements if not pure.
```

Comment:
The comments seem to be slightly off here, I assume they would be one single comment?

Reply:
Or, even better: just use `self.ket()` and `self.dm()` directly inside the if statements below, to avoid creating a `state` variable that could represent either.

Reply:
Why should I use methods instead of grabbing directly the data? Seems like wanting to go through the middle man instead of getting directly to the substance.

Reply:
`self.ket()` and `self.dm()` simply return the `self.data` directly. However it's

* More future proof (changes to how the data is stored means we just need to update `self.ket()` and `self.dm()`, not `number_expectation` as well)

* Leads to cleaner code